p2p/monero-p2p/tests/fragmented_handshake.rs

```diff
@@ -0,0 +1,224 @@
+    network_zones::ClearNetServerCfg,
+    ConnectionDirection, NetworkZone,
+};
+use monero_wire::{
+    common::PeerSupportFlags,
+    levin::{message::make_fragmented_messages, LevinMessage, Protocol},
+    BasicNodeData, Message, MoneroWireCodec,
+};
+
+use cuprate_test_utils::monerod::monerod;
+
+mod utils;
+use utils::*;
+
+/// A network zone equal to clear net where ever message sent is turned into a fragmented message.
+/// Does not support sending fragmented or dummy messages manually.
+#[derive(Clone, Copy)]
+pub enum FragNet {}
+
+#[async_trait::async_trait]
+impl NetworkZone for FragNet {
+    const NAME: &'static str = "FragNet";
+    const ALLOW_SYNC: bool = true;
+    const DANDELION_PP: bool = true;
+    const CHECK_NODE_ID: bool = true;
+
+    type Addr = SocketAddr;
+    type Stream = FramedRead<OwnedReadHalf, MoneroWireCodec>;
+    type Sink = FramedWrite<OwnedWriteHalf, FragmentCodec>;
+    type Listener = InBoundStream;
+
+    type ServerCfg = ClearNetServerCfg;
+
+    async fn connect_to_peer(
+        addr: Self::Addr,
+    ) -> Result<(Self::Stream, Self::Sink), std::io::Error> {
+        let (read, write) = TcpStream::connect(addr).await?.into_split();
+        Ok((
+            FramedRead::new(read, MoneroWireCodec::default()),
+            FramedWrite::new(write, FragmentCodec::default()),
+        ))
+    }
+
+    async fn incoming_connection_listener(
+        config: Self::ServerCfg,
+    ) -> Result<Self::Listener, std::io::Error> {
+        let listener = TcpListener::bind(config.addr).await?;
+        Ok(InBoundStream { listener })
+    }
+}
+
+pub struct InBoundStream {
+    listener: TcpListener,
+}
+
+impl Stream for InBoundStream {
+    type Item = Result<
+        (
+            Option<SocketAddr>,
+            FramedRead<OwnedReadHalf, MoneroWireCodec>,
+            FramedWrite<OwnedWriteHalf, FragmentCodec>,
+        ),
+        std::io::Error,
+    >;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.listener
+            .poll_accept(cx)
+            .map_ok(|(stream, addr)| {
+                let (read, write) = stream.into_split();
+                (
+                    Some(addr),
+                    FramedRead::new(read, MoneroWireCodec::default()),
+                    FramedWrite::new(write, FragmentCodec::default()),
+                )
+            })
+            .map(Some)
+    }
+}
+
+#[derive(Default)]
+pub struct FragmentCodec(MoneroWireCodec);
+
+impl Encoder<LevinMessage<Message>> for FragmentCodec {
+    type Error = <MoneroWireCodec as Encoder<LevinMessage<Message>>>::Error;
+
+    fn encode(
+        &mut self,
+        item: LevinMessage<Message>,
+        dst: &mut BytesMut,
+    ) -> Result<(), Self::Error> {
+        match item {
+            LevinMessage::Body(body) => {
+                // 66 is the minimum fragment size.
+                let fragments = make_fragmented_messages(&Protocol::default(), 66, body).unwrap();
```

Comment:
Is this/should this be in a const somewhere?

Reply:
This is just a test and no one should be using fragments this small for real messages, monerod uses 3_000.